proposals/004_Overlays.md

```diff
@@ -0,0 +1,194 @@
+      info:
+        x-overlay-applied: structured-overlay
+      paths:
+        "/":
+          summary: "The root resource"
+          get:
+            summary: "Retrieve the root resource"
+            x-rate-limit: 100
+        "/pets":
+          get:
+            summary: "Retrieve a list of pets"
+            x-rate-limit: 100
+      components:
+      tags:
+```
+
+##### Targeted Overlays
+
+Alternatively, where only a small number of updates need to be applied to a large document, each [Update Object](#updateObject) can be more targeted.
+
+```yaml
+overlay: 1.0.0
+info:
+  title: Structured Overlay
+  version: 1.0.0
+updates:
+- target: paths."/foo".get
+    value:
+        description: This is the new description
+- target: paths."/bar".get
+    value:
+        description: This is the updated description
+- target: paths."/bar"
+    value:
+        post:
+            description: This is an updated description of a child object
+            x-safe: false
+```
+
+##### Wildcard Overlays Examples
+
+One significant advantage of using the JMESPath syntax that it allows referencing multiple nodes in the target document.  This would allow a single update object to be applied to multiple target objects using wildcards.
+
+```yaml
+overlay: 1.0.0
+info:
+  title: Update many objects at once
+  version: 1.0.0
+updates:
+- target: paths.*.get
+    value:
+      x-safe: true
+- target: paths.*.get.parameters[?name=='filter' && in=='query']
+    value:
+      schema:
+        $ref: "/components/schemas/filterSchema"
+```
+
+##### Array Modification Examples
+
+Due to the fact that we can now reference specific elements of the parameter array, it allows adding parameters. Parameters can be deleted using the `remove` property.  Use of indexes to remove array items should be avoided where possible as indexes will change when items are removed.
+
+```yaml
+overlay: 1.0.0
+info:
+  title: Add an array element
+  version: 1.0.0
+updates:
+- target: paths.*.get.parameters[length(@)]
+    value: 
+      name: newParam
+      in: query
+```
+
+```yaml
+overlay: 1.0.0
+info:
+  title: Remove a array element
+  version: 1.0.0
+updates:
+- target: $.paths[*].get.parameters[? name == 'dummy']
+    remove: true
+```
+
+## Proposal Summary
+
+### Benefits
+
+- This approach addresses the two distinct approaches of structured overlay vs targeted overlay which suits distinct but equally valid scenarios.
+- Addresses the problem of modifying the parameters array and removes the need to replace the entire array when a small change is required.
+- Allows sets of related overlays to be stored in a same file.
+- Enables updating a set of objects based on a pattern. This might be an effective way of apply common behaviour across many operations in an API.
+
+### Challenges
+
+- Tooling will need a JMESPath implementation.
+- Large overlays may be slow to process.
+- Multiple complex pattern based overlays may cause overlapping updates causing confusing outcomes.
```

Comment:
The overlay order needs to be unambiguous. Moreover, previous overlays in a chain must be immutable. A change to an earlier overlay will likely ruin subsequent overlays. Perhaps an overlay can declare a hash of the document it is updating? This would probably need to be coupled with some kind of ordering convention in the overlay files.

Reply:
That last comment of mine is poorly worded.   An Overlay document has an array of update objects.  The order of applying those updates is defined based on the order of the array.  Therefore within an Overlay document the order is well defined.  However, there still could be overlapping updates.  That could make the Overlay document tricky to understand.  Best practice should be to try and avoid the situation.

Having multiple Overlay documents apply to a single OpenAPI description is a different challenge.  Currently this is up to tooling to manage the order of application.  I don't like the idea of linking or ordering Overlay documents explicitly in the document, but I could imagine the use of priority value that provides an ordering hint to tooling.